### ISSUE TITLE
Symbol From Discarded Section Referenced

### ISSUE THREAD

**Issue by Grant12311** [skypjack/entt#633]
Hello. I've recently been working on implementing entt in my game engine. I've run into a very strange and unfortunate issue and I would really appreciate some help. While I'm not sure my problem lies in entt's code, it's starting to look that way. Below is the linker error that is generated by the default GCC linker.
```
`_ZZN4entt8internal9type_nameIN4GJGO20Transform2DComponentEEEDavE19__PRETTY_FUNCTION__' referenced in section `.data' of ../../bin/LinuxARM/Debug/libGJGO.a(scene_management_layer.o): defined in discarded section `.rodata._ZZN4entt8internal9type_nameIN4GJGO20Transform2DComponentEEEDavE19__PRETTY_FUNCTION__[_ZZN4entt9type_nameIN4GJGO20Transform2DComponentEvE5valueEvE5value]' of ../../bin/LinuxARM/Debug/libGJGO.a(scene_management_layer.o)
```

Below is the error given by GCC's "gold" linker.
```
../../bin/LinuxARM/Debug/libGJGO.a(scene_management_layer.o)(.data+0x208): error: relocation refers to local symbol "_ZZN4entt8internal9type_nameIN4GJGO20Transform2DComponentEEEDavE19__PRETTY_FUNCTION__" [998], which is defined in a discarded section
  section group signature: "_ZZN4entt9type_nameIN4GJGO20Transform2DComponentEvE5valueEvE5value"
  prevailing definition is from obj/LinuxARM/Debug/src/main.o
collect2: error: ld returned 1 exit status
```

Below is the error report given by my address sanitizer (Asan, I believe).
```
AddressSanitizer:DEADLYSIGNAL
=================================================================
==20540==ERROR: AddressSanitizer: SEGV on unknown address 0x00000000 (pc 0x00000000 bp 0xbeeb663c sp 0xbeeb65a0 T0)
==20540==Hint: pc points to the zero page.
==20540==The signal is caused by a READ memory access.
==20540==Hint: address points to the zero page.

AddressSanitizer can not provide additional info.
SUMMARY: AddressSanitizer: SEGV (<unknown module>) 
==20540==ABORTING
```

I should note that using G++ 8.3.0, it would not link (i. e. it gave an actual fatal error, not a warning). But, when I compiled the main file with G++ 10 (I don't have G++ 10 install, I was using a hastily installed binary, which is why I didn't build the entire project with it. I couldn't integrate it with my build system) and then link it with G++ 8.3.0 - which I had to do because the G++ 10 linker wasn't working for me - it would link "successfully" and output a binary. Running it gave the sanitizer report. 

Below is a stack trace obtained from GDB.
```
Program received signal SIGSEGV, Segmentation fault.
0x00000000 in ?? ()
(gdb) bt
#0  0x00000000 in ?? ()
#1  0x0012375c in entt::swap (lhs=..., rhs=...)
    at /home/pi/C++/GJGO/vendor/entt/src/entt/entity/../poly/../core/any.hpp:278
#2  0x00137ee0 in entt::swap (lhs=..., rhs=...)
    at /home/pi/C++/GJGO/vendor/entt/src/entt/entity/../poly/poly.hpp:311
#3  0x0014b7ec in entt::poly<entt::Storage<entt::entity> >::poly (
    this=0xb1a10988, other=...)
    at /home/pi/C++/GJGO/vendor/entt/src/entt/entity/../poly/poly.hpp:237
#4  0x0014aea8 in entt::basic_registry<entt::entity>::pool_data::pool_data (
    this=0xb1a10988)
    at /home/pi/C++/GJGO/vendor/entt/src/entt/entity/registry.hpp:51
#5  0x0014af20 in __gnu_cxx::new_allocator<entt::basic_registry<entt::entity>::pool_data>::construct<entt::basic_registry<entt::entity>::pool_data, entt::basic_registry<entt::entity>::pool_data> (this=0xb306b550, __p=0xb1a10988)
    at /opt/gcc-10.1.0/include/c++/10.1.0/ext/new_allocator.h:150
#6  0x0014a0b0 in std::allocator_traits<std::allocator<entt::basic_registry<entt::entity>::pool_data> >::construct<entt::basic_registry<entt::entity>::pool_data, entt::basic_registry<entt::entity>::pool_data> (__a=..., __p=0xb1a10988)
    at /opt/gcc-10.1.0/include/c++/10.1.0/bits/alloc_traits.h:512
#7  0x00148cd0 in std::__relocate_object_a<entt::basic_registry<entt::entity>::pool_data, entt::basic_registry<entt::entity>::pool_data, std::allocator<entt::basic_registry<entt::entity>::pool_data> > (__dest=0xb1a10988, 
--Type <RET> for more, q to quit, c to continue without paging--
    __orig=0xb24a4bc8, __alloc=...)
    at /opt/gcc-10.1.0/include/c++/10.1.0/bits/stl_uninitialized.h:976
#8  0x00146bcc in std::__relocate_a_1<entt::basic_registry<entt::entity>::pool_data*, entt::basic_registry<entt::entity>::pool_data*, std::allocator<entt::basic_registry<entt::entity>::pool_data> > (__first=0xb24a4bc8, __last=0xb24a4bf0, 
    __result=0xb1a10960, __alloc=...)
    at /opt/gcc-10.1.0/include/c++/10.1.0/bits/stl_uninitialized.h:1014
#9  0x0014493c in std::__relocate_a<entt::basic_registry<entt::entity>::pool_data*, entt::basic_registry<entt::entity>::pool_data*, std::allocator<entt::basic_registry<entt::entity>::pool_data> > (__first=0xb24a4ba0, __last=0xb24a4bf0, 
    __result=0xb1a10960, __alloc=...)
    at /opt/gcc-10.1.0/include/c++/10.1.0/bits/stl_uninitialized.h:1028
#10 0x00141d64 in std::vector<entt::basic_registry<entt::entity>::pool_data, std::allocator<entt::basic_registry<entt::entity>::pool_data> >::_S_do_relocate (
    __first=0xb24a4ba0, __last=0xb24a4bf0, __result=0xb1a10960, __alloc=...)
    at /opt/gcc-10.1.0/include/c++/10.1.0/bits/stl_vector.h:456
#11 0x0013cc40 in std::vector<entt::basic_registry<entt::entity>::pool_data, std::allocator<entt::basic_registry<entt::entity>::pool_data> >::_S_relocate (
    __first=0xb24a4ba0, __last=0xb24a4bf0, __result=0xb1a10960, __alloc=...)
    at /opt/gcc-10.1.0/include/c++/10.1.0/bits/stl_vector.h:469
#12 0x00136b80 in std::vector<entt::basic_registry<entt::entity>::pool_data, std::allocator<entt::basic_registry<entt::entity>::pool_data> >::_M_default_append
    (this=0xb306b550, __n=1)
--Type <RET> for more, q to quit, c to continue without paging--
    at /opt/gcc-10.1.0/include/c++/10.1.0/bits/vector.tcc:648
#13 0x0013066c in std::vector<entt::basic_registry<entt::entity>::pool_data, std::allocator<entt::basic_registry<entt::entity>::pool_data> >::resize (
    this=0xb306b550, __new_size=3)
    at /opt/gcc-10.1.0/include/c++/10.1.0/bits/stl_vector.h:940
#14 0x001c0d70 in entt::basic_registry<entt::entity>::assure<GJGO::SpriteComponent> (this=0xb306b544) at vendor/entt/src/entt/entity/registry.hpp:111
#15 0x001bee4c in entt::basic_registry<entt::entity>::assure<GJGO::SpriteComponent> (this=0xb306b544) at vendor/entt/src/entt/entity/registry.hpp:127
#16 0x001be244 in entt::basic_registry<entt::entity>::view<GJGO::Transform2DComponent, GJGO::SpriteComponent>(entt::exclude_t<>) (this=0xb306b544)
    at vendor/entt/src/entt/entity/registry.hpp:1138
#17 0x001bd644 in GJGO::SceneManagementLayer::draw (this=0xb3e21d10)
    at /home/pi/C++/GJGO/src/scene_management_layer.cpp:24
#18 0x0019ef98 in GJGO::Application::run (this=0xbefff060)
    at /home/pi/C++/GJGO/src/app.cpp:103
#19 0x00122674 in main ()
    at /home/pi/C++/GJGO/examples/Sprite Test/src/main.cpp:231
```

My complete code is available in [this](https://github.com/Grant12311/GJGO-Engine/tree/entity) GitHub repository *in the "entity" branch only*. All issues have come from the Sprite Test program in the examples folder. It is also worth noting that the application links in Release and Dist builds, though it crashes with a segmentation fault when I try to run it. Forgive the lack of documentation. This code was never meant to be public. By removing code until the project linked, I was able to trace the issue down to line 27 of "scene_management_layer.cpp" in the src folder (the line that obtains the entt::view for iteration). It is noteworthy that the above stack trace points to a slightly different line in that file. It should also be noted that this issue only appears if one or more components have been added to the "dinoEntity" entity in GameLayer's constructor in Sprite Test's main.cpp file. GJGO::Entity is my entt abstraction. It is a rather straight forward abstraction and can be found in entity.hpp and entity.cpp. I should also explain where the registry is located. Registries exist in GJGO::Scene objects. One is created in the GJGO::Application class's constructor. Entities are created from a member function of that scene.

Now a note on compatibility. My project uses Premake5 as its build system. Most dependencies are sub-modules located in the vendor folder, though a few things aren’t (GLFW being one). I’ve only ever used my project on Linux, so I can’t be sure it’ll work on other operating systems. I know for a fact that clip, the library I use for clipboard functionality, is being compiled for Linux only. However, this should be easy to  remedy by editing the project.lua premake file in the clip sub-module to include files for other OS(s). Alternatively, you could just disable the code in the clipboard source files and stop linking/including clip at all, as it shouldn't be necessary. Lastly, we have graphics API support. GLES3 is the only target I’ve ever had. Because of this, things like GLFW context creation hints and shader version strings are hard-coded for GLES3. Thankfully, I expect it to be trivial to adapt the code for desktop OpenGL. Please contact me if you run into any issues.

One final note. I have a friend who is 100% convinced that undefined behavior exists in a function somewhere that is causing the compiler to discard said function. The linker then fails when it tries to find the discarded function. I personally have no idea if this is the case. I just found it worth mentioning.

That you very much for your time. Sorry if I did something wrong, as I'm relatively new at working with GitHub's issue system. Also, don't hesitate to ask me for more information. I'll gladly answer any questions pertaining to my source code. I understand that it is very poorly documented and probably not very easy for third parties to navigate. Have a wonderful day.

**Comment by skypjack** [comment]
I've been trying to figure out what the root cause of this could be for at least a couple of days, but I've no clue yet. This is the first time that I see this error and to be honest I don't even know how to reproduce it. Is there any chance that you can provide a minimal example to reproduce the error? A whole project full of unrelated things and that works only on linux isn't exaclty the best tool for the purpose. 🙂 
Btw I would ignore the case in which you compiled with a compiler version and linked with an older one. This sounds somewhat _wrong_ anyway, despite the error. 😄 
To start with, can you provide the compilation line and therefore all the options passed to the compiler upon compilation?

**Comment by Grant12311** [comment]
I have a new interesting development. A friend just complied my code on his hardware and everything worked fine. He has a slightly newer GCC version than I do (8.3 vs. 8.4). I’m going to try using a newer compiler and report back my findings. If that doesn’t help, I’ll get the info you requested. Though I suspect there really is nothing wrong with anyone’s code if it worked for him.

**Comment by skypjack** [comment]
Well, I would also suggest to clean your build dir before doing anything else.
I'm looking forward to your updates. :+1:
Thanks for reporting.

**Comment by Grant12311** [comment]
I just wanted to give you a quick update. Shortly after I last talked, my USB controller drivers decided to stop working for no apparent reason. That was a significant delay, since the drives are part of the Linux kernel. Now that I’ve got that sorted, I’m compiling a newer version of GCC than the one available in my package manager. Specifically, I’m compiling the version that my aforementioned friend used to successfully build my project. Once that’s done, I’ll report back if that solved the problem. If it does, I’d assume it was a bug with GCC 8.3. If not, I’ll get to work on that simplified example you requested.

**Comment by Grant12311** [comment]
I can now confirm that the problem still exists with GCC 8.4. Also, I cloned the repo into a completely new location to endure a clean build directory.

**Comment by skypjack** [comment]
It's interesting, because this is the first time someone reports this and honestly also the first time I've ever heard of this error in general.
If you could help to reproduce it, I'd be interested in looking into your issue.

**Comment by Grant12311** [comment]
As far as I can tell, the issue has now been resolved. Sadly, I’m not fully aware of what I did to fix it, but everything seems to work now. It appears to have been an issue with how my project’s Entt submodule was setup. This caused Entt to improperly clone in some circumstances. After removing and re-adding Entt, all issues seem to have been resolved. Thank you very much for you time. I apologize for getting you involved when it wasn’t even a problem with your code.
